src/platform/backends/qemu/qemu_virtual_machine_factory.cpp

```diff
@@ -59,6 +89,72 @@ mp::VirtualMachine::UPtr mp::QemuVirtualMachineFactory::create_virtual_machine(c
                                                     get_instance_directory(desc.vm_name));
 }
 
+mp::VirtualMachine::UPtr mp::QemuVirtualMachineFactory::create_vm_and_clone_instance_dir_data(
+    const VMSpecs& /*src_vm_spec*/,
+    const VMSpecs& dest_vm_spec,
+    const std::string& source_name,
+    const std::string& destination_name,
+    const VMImage& dest_vm_image,
+    const SSHKeyProvider& key_provider,
+    VMStatusMonitor& monitor)
+{
+    const std::filesystem::path source_instance_data_directory{get_instance_directory(source_name).toStdString()};
+    const std::filesystem::path dest_instance_data_directory{get_instance_directory(destination_name).toStdString()};
+
+    // if any of the below code throw, then roll back and clean up the created instance folder
+    auto rollback_delete_instance_folder =
+        sg::make_scope_guard([dest_instance_directory = dest_instance_data_directory]() noexcept -> void {
+            // use err_code to guarantee the two file operations below do not throw
+            if (std::error_code err_code; MP_FILEOPS.exists(dest_instance_directory, err_code))
+            {
+                fs::remove_all(dest_instance_directory, err_code);
+                if (err_code.value())
+                {
+                    mpl::log(
+                        mpl::Level::info,
+                        category,
+                        fmt::format("The rollback instance directory removal did not succeed, err_code message is : {}",
+                                    err_code.message()));
+                }
+            }
+        });
+
+    copy_instance_dir_without_snapshot_files(source_instance_data_directory, dest_instance_data_directory);
+
+    const fs::path cloud_init_config_iso_file_path = dest_instance_data_directory / "cloud-init-config.iso";
+
+    MP_CLOUD_INIT_FILE_OPS.update_cloned_cloud_init_unique_identifiers(dest_vm_spec.default_mac_address,
+                                                                       dest_vm_spec.extra_interfaces,
+                                                                       destination_name,
+                                                                       cloud_init_config_iso_file_path);
+
+    // start to construct VirtualMachineDescription
+    mp::VirtualMachineDescription dest_vm_desc{dest_vm_spec.num_cores,
```

Comment:
Do you mean creating `dest_vm_desc` in `daemon::clone` and passing it to `create_vm_and_clone_instance_dir_data` function? That would make the `create_vm_and_clone_instance_dir_data`'s parameter list more cumbersome since the parameters like dest_vm_spec, destination_name are needed by the function without creating `dest_vm_desc`.

Reply:
Ok, I think it's fine the way it is. This way the daemon is agnostic.